### PR TITLE
WIP - add RowRange object

### LINKED ISSUE #3308: OnDemand Follow-on: Create a public RowRange object with start/end rows and exclusivity booleans
Reference comment at: https://github.com/apache/accumulo/pull/3294#discussion_r1167115920. There are a lot of places in the API where `Text start` and `Text end` can be set by the user to specify a Range. This task is about creating a common object that captures the start, end, and inclusivity of both, and then using that in all places in the API where a range is being passed.

### PR DESCRIPTION
**Pull request by DomGarguilo**
Closes #3308

### CODE REVIEW

File: core/src/main/java/org/apache/accumulo/core/data/RowRange.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+import org.apache.hadoop.io.Text;
+
+public class RowRange {
+  final private Text startRow;
+  final private boolean startInclusive;
+  final private Text endRow;
+  final private boolean endInclusive;
+
+  public RowRange(Text startRow, Text endRow) {
```
Comment (dlmarion) [inline core/src/main/java/org/apache/accumulo/core/data/RowRange.java:29]: I think we also want a no-arg constructor like in `Range` to convey +/- infinity and with that the `isInfiniteKey` methods

File: core/src/main/java/org/apache/accumulo/core/client/admin/FindMax.java [hunk 2]
```diff
@@ -181,4 +180,9 @@ public static Text findMax(Scanner scanner, Text start, boolean is, Text end, bo
 
     return _findMax(scanner, start, is, end, ie);
   }
+
+  public static Text findMax(Scanner scanner, RowRange rowRange) {
```
Comment (dlmarion) [inline core/src/main/java/org/apache/accumulo/core/client/admin/FindMax.java:184]: I think you just want this one `findMax` method that uses the RowRange parameter.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java [hunk 3]
```diff
@@ -1560,6 +1562,15 @@ public Text getMaxRow(String tableName, Authorizations auths, Text startRow,
     return FindMax.findMax(scanner, startRow, startInclusive, endRow, endInclusive);
   }
 
+  @Override
+  public Text getMaxRow(String tableName, Authorizations auths, RowRange rowRange)
```
Comment (dlmarion) [inline core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java:1566]: The other getMaxRow method should call this getMaxRow method, creating a RowRange from the args.

**Comment by cshannon** [comment]
Having a RowRange object would be useful in 3.1 and not just elasticity. As part of #1327 I need to use a range for fencing off table files. Right now in my prototypes I'm just using the Range object and leaving cq/cf, etc null/empty but really we just need to fence off by rows.

It would need to support some of the original methods though of Range for it to be used. There are some very useful methods in Range like [mergeOverlapping](https://github.com/cshannon/accumulo/blob/9a31c4257b54ed890f02ab9aafe6d99a9638dac6/core/src/main/java/org/apache/accumulo/core/data/Range.java#L418) and [clip](https://github.com/cshannon/accumulo/blob/9a31c4257b54ed890f02ab9aafe6d99a9638dac6/core/src/main/java/org/apache/accumulo/core/data/Range.java#L511) that would need to exist.

After I talked about this with @keith-turner last week I was thinking a one approach would be to do some refactoring and move things just for rows into RowRange and then have Range extend RowRange and add CF, CQ, etc.

**Comment by DomGarguilo** [comment]
@cshannon, interesting ideas. I am planning to move these changes to target 3.1. I will also look into the other ideas as I go.

**Comment by cshannon** [comment]
> @cshannon, interesting ideas. I am planning to move these changes to target 3.1. I will also look into the other ideas as I go.

Sounds good, there's no rush. I can and will keep using Range in my no-chop merge prototype for now as it would be easy to swap it out if/when a RowRange object is ready if it supported all the utility methods as I said. I can also review this too and look more at it to help out. 

In theory this object should be able to be re-worked from the existing Range object and support the existing methods (even if it's just a shared interface or something and not inheritance) because supporting only rows should be simpler logic to implement than the existing object that handles an entire Key.

File: core/src/main/java/org/apache/accumulo/core/data/RowRange.java (hunk 1 above)
Comment (DomGarguilo) [reply core/src/main/java/org/apache/accumulo/core/data/RowRange.java:48]: addressed in 49c98bd

File: core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java (hunk 3 above)
Comment (DomGarguilo) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java:1564]: Addressed in 3a6d615

**Comment by DomGarguilo** [comment]
Superseded by #3342

**Pull request closed without merging**